feat: (Core) introduce new Carousel component

DeepakSap14 opened the pull request:
#### Please provide a link to the associated issue.
https://github.com/SAP/fundamental-ngx/issues/3279

#### Please provide a brief summary of this pull request.
Introducing carousel component in core.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] update `README.md`
- [x] Documentation Examples
- [x] Stackblitz works for all examples

Known issues:
1. Looping is not working for Multiple active item.
2. Mobile swipe functionality not implemented.

sKudum commented on libs/core/src/lib/carousel/carousel.component.ts:
```diff
@@ -0,0 +1,309 @@
+})
+export class CarouselComponent implements OnInit, AfterContentInit, OnDestroy {
+    /** @hidden */
+    @ContentChildren(CarouselItemComponent)
```
we need to follow angular coding guidelines for properties alignment. reference angular.io,  we can update where ever it is applicable

DeepakSap14 replied:
fixed. followed angular guidelines.

sKudum commented on libs/core/src/lib/carousel/carousel.component.ts:
```diff
@@ -0,0 +1,309 @@
+
+    /** Id for the Carousel. */
+    @Input()
+    @HostBinding('attr.id')
```
It is recommended not to use @hostbinding()  along with @input()

DeepakSap14 replied:
I see its getting used in core components. Do you have any other approach?

JKMarkowski replied:
IMO IT's correct

sKudum commented on libs/core/src/lib/carousel/carousel-item/carousel-item.component.ts:
```diff
@@ -0,0 +1,77 @@
+    changeDetection: ChangeDetectionStrategy.OnPush,
+    encapsulation: ViewEncapsulation.None
+})
+export class CarouselItemComponent {
```
Base component extension won't help here?

DeepakSap14 replied:
Base component is in platform. carousel is in core. so will not be able to use base component.

sKudum commented on libs/core/src/lib/carousel/carousel-item/carousel-item.component.spec.ts:
```diff
@@ -0,0 +1,25 @@
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+});
```
having few test scenarios will help.

DeepakSap14 replied:
have included tests in carousel component for carousel items as well. removed this file.

sKudum commented on libs/core/src/lib/carousel/carousel.module.ts:
```diff
@@ -0,0 +1,11 @@
+    imports: [CommonModule],
+    exports: [CarouselComponent, CarouselItemComponent]
+})
+export class CarouselNewModule {}
```
Only core we have a carousel why should we have "New"

DeepakSap14 replied:
using CarouselModule now.

mikerodonnell89 commented:
The images in the preview take a very long time to load if they're not cached, perhaps each example can share a service that will load the images so they'll only need to be downloaded once. Also I think that service itself is just very slow so maybe use a different service, placeimg seems to be faster

mikerodonnell89 commented:
Thumb swiping for the mobile example (hidden navigation buttons) does not work

mikerodonnell89 commented:
Playing with the width/height inputs, the carousel items overflow beyond the boundaries, is that intentional?

mikerodonnell89 commented:
I think `<div style="display: flex; justify-content: center; margin-top: 0px;">` can be built in to the component rather than requiring the dev to add it themselves

droshev reviewed (changes requested):
have you implemented the sliding effect:
```
When the user navigates from the current item to another item, the current item is moved out of the content area, and the next or previous item slides in (depending on the direction of navigation).

On touch devices, users navigate with swipe gestures (swipe right or swipe left).
```
also the animation is missing when you change the component in

droshev commented:
> I think `<div style="display: flex; justify-content: center; margin-top: 0px;">` can be built in to the component rather than requiring the dev to add it themselves

I agree with @mikerodonnell89 and also it doesn't follow the markup from styles

droshev commented:
> ork for coz

the sliding should work on any device including the browser

kavya-b commented on apps/docs/src/app/core/component-docs/carousel/examples/carousel-content-navigation.component.html:
```diff
@@ -0,0 +1,39 @@
+<fd-carousel [navigatorInPageIndicator]="false" [pageIndicatorContainerPlacement]="'top'">
```
remove `"'top'"` [] binding here and at other places

DeepakSap14 replied:
fixed

kavya-b commented on libs/core/src/lib/carousel/carousel.component.ts:
```diff
@@ -0,0 +1,315 @@
+})
+export class CarouselComponent implements OnInit, AfterContentInit, OnDestroy {
+    /** @hidden */
+    @ContentChildren(CarouselItemComponent, { descendants: true })
```
I think `@ContentChildren` is supposed to come after `Input()` and `Output()`, right?

DeepakSap14 replied:
fixed

kavya-b commented on libs/core/src/lib/carousel/carousel.component.ts:
```diff
@@ -0,0 +1,315 @@
+const ICON_PAGE_INDICATOR_LIMIT = 8;
+let carouselUniqueId = 0;
+
+class CarouselActiveItem {
```
would this need to be exported?

DeepakSap14 replied:
fixed. export added

Betrozov commented on apps/docs/src/app/core/component-docs/carousel/examples/carousel-no-page-indicator.component.html:
```diff
@@ -0,0 +1,35 @@
+<fd-carousel [showPageIndicator]="false" [pageIndicatorContainerPlacement]="'top'">
```
unnecessary binding

DeepakSap14 replied:
fixed

JKMarkowski commented:
I agree with @droshev, this should be used with some functionality of `carousel.directive` from `utils`. There is already implemented swipe support

JKMarkowski commented:
@DeepakSap14 I preapred some service, by copy/paste some properties from `carousel.directive`. Check it there: https://github.com/SAP/fundamental-ngx/blob/fix/create-carousel-service/libs/core/src/lib/utils/services/carousel.service.ts

There you can see how it can be used:
https://github.com/SAP/fundamental-ngx/blob/fix/create-carousel-service/libs/core/src/lib/utils/directives/carousel/carousel.directive.ts

KevinOkamoto reviewed (approved):
Looks good.

DeepakSap14 commented:
Now i have included Sliding effect for browsers and swiping effect for touch devices as well.
i have included example where cards can be used as carousel item as well.
please review now.

DeepakSap14 commented:
> The images in the preview take a very long time to load if they're not cached, perhaps each example can share a service that will load the images so they'll only need to be downloaded once. Also I think that service itself is just very slow so maybe use a different service, placeimg seems to be faster

Creating service to fetch these images gives Cross-origin error.
I have changed the image provider to "placeimg", hope this will be good enough.

mykolavorobiov commented on libs/core/src/lib/utils/services/carousel.service.ts:
```diff
@@ -0,0 +1,297 @@
+    private _hammerSetup(): void {
+        this._hammer = new Hammer(this._elementRef.nativeElement);
+
+        this._hammer.get('pan').set({ direction: Hammer.DIRECTION_ALL });
```
Could please apply direction depends on your config? for example, `this.config.vertical` should use only `Hammer.DIRECTION_VERTICAL`

DeepakSap14 replied:
@JKMarkowski , do you want me to make this change. this code was present from earlier. so please confirm.

DeepakSap14 replied:
changed to Hammer.DIRECTION_VERTICAL inside this.config.vertical condition.
fixed.

mykolavorobiov commented on libs/core/src/lib/carousel/carousel.component.ts:
```diff
@@ -0,0 +1,404 @@
+                this.rightButtonDisabled = true;
+            }
+
+            if (this.currentActiveSlidesStartIndex > 0) {
```
Could you please move `this.leftButtonDisabled = false;` and `this.rightButtonDisabled = false;` to above construction with `...else...`, and remove unnecessary conditions

`if (this.currentActiveSlidesStartIndex === 0) {`
`                this.leftButtonDisabled = true;`
`            } else if (this.currentActiveSlidesStartIndex === this.slides.length - 1) {`
`                this.rightButtonDisabled = true;`
`            } else {`
`this.leftButtonDisabled = false;`
`this.rightButtonDisabled = false;`
`}`

DeepakSap14 replied:
fixed.

mykolavorobiov commented on libs/core/src/lib/carousel/carousel.component.ts:
```diff
@@ -0,0 +1,404 @@
+                this.leftButtonDisabled = true;
+                this.rightButtonDisabled = true;
+            }
+            if (this.currentActiveSlidesStartIndex === 0) {
```
Looks like the same as the above conditions.

DeepakSap14 replied:
fixed.

mykolavorobiov reviewed (changes requested):
Looks good, please take a look few minor comments

JKMarkowski reviewed (changes requested):
Hi @DeepakSap14, good job, most of examples work great, but

![carousel](https://user-images.githubusercontent.com/26483208/95985196-9832a700-0e24-11eb-94b7-499d525fda4e.gif)

As you can see slide doesn't work good, indicator doesn't stop on last element. 
Also TimeComponent doesn't work with those changes
![image](https://user-images.githubusercontent.com/26483208/95985308-bbf5ed00-0e24-11eb-8d67-2274a8b7444b.png)

DeepakSap14 commented:
> Hi @DeepakSap14, good job, most of examples work great, but
> 
> ![carousel](https://user-images.githubusercontent.com/26483208/95985196-9832a700-0e24-11eb-94b7-499d525fda4e.gif)
> 
> As you can see slide doesn't work good, indicator doesn't stop on last element.
> Also TimeComponent doesn't work with those changes
> ![image](https://user-images.githubusercontent.com/26483208/95985308-bbf5ed00-0e24-11eb-8d67-2274a8b7444b.png)

1. Fixed, extra swiping happening in carousel service in case of multiple item display.
2. Time component is using "carousel directive" and internally it is using "fd-carousel" class.
     so namespace conflict is happening for "fd-carousel", when page is switched from "carousel" to "time" component.
    I tried directly importing "fundamental-styles/carousel.scss" in time-column.component.scss but this also does not resolve 
    issue.
   **As a temporary fix**: i have disabled "fd-carousel" class load on "fdCarouselDirective" by appending extra characted to class name.
  This does not introduce any issue in time component examples, all examples works good in time component.
  If needed we have to raise a separate issue and provide fix in "time component".

DeepakSap14 commented:
@JKMarkowski please review this.

JKMarkowski commented:
Hi @DeepakSap14 I need to ask you how the API there. I see multi item doesn't have any `[swipeEnabled]` and the swipe works fine (on desktop). For mobile swipe works everywhere and it doesn't matter if `[swipeEnabled]` is set to true

JKMarkowski commented:
![carouselloop](https://user-images.githubusercontent.com/26483208/96264281-4fb1ef80-0fc4-11eb-8a54-2bff6f2b0903.gif)
After enabling swipe on loop carousel, it's a little bit bugged

droshev commented:
> ![carouselloop](https://user-images.githubusercontent.com/26483208/96264281-4fb1ef80-0fc4-11eb-8a54-2bff6f2b0903.gif)
> After enabling swipe on loop carousel, it's a little bit bugged

Can we enable this with mouse down + swipe left/right?

DeepakSap14 commented:
> > ![carouselloop](https://user-images.githubusercontent.com/26483208/96264281-4fb1ef80-0fc4-11eb-8a54-2bff6f2b0903.gif)
> > After enabling swipe on loop carousel, it's a little bit bugged
> 
> Can we enable this with mouse down + swipe left/right?

swipe event is coming from hammerjs. handling this with key will not be possible.
we have left and right arrow navigation alrerady.

DeepakSap14 commented:
@JKMarkowski fixed bug for looped carousel. please check now.

DeepakSap14 commented:
> Hi @DeepakSap14, Great Job with it! I just need to ask you how the API there. I see [multi item example](https://deploy-preview-3551--fundamental-ngx.netlify.app/#/core/carousel#carousel-multi-item-active) doesn't have any `[swipeEnabled]` and the swipe works fine (on desktop). For mobile swipe works fine everywhere.

By default i have made swipeEnabled="true", so it will be work in touch devices without any change.

DeepakSap14 commented:
@JKMarkowski @droshev Please review and approve.
It has passed expected due date. If some minor issue, i will raise a different issue to track.

sKudum reviewed (approved):
My comments are addressed

dimamarksman commented on libs/core/src/lib/fundamental-ngx-core.module.ts:
```diff
@@ -142,6 +143,6 @@ import { MessageToastService } from './message-toast/message-toast-service/messa
         MessageToastModule
     ],
-    providers: [AlertService, DialogService, NotificationService, MessageToastService]
+    providers: [AlertService, CarouselService, DialogService, NotificationService, MessageToastService]
```
We don't need to include `CarouselDirective` since it's defined as provider on the CarouselDirective level.
Please correct me if I'm wrong.

DeepakSap14 replied:
fixed now. Does not need service here.

dimamarksman reviewed (approved):
Looks good!
Just a few questions

JKMarkowski commented:
IE11
![ie11carousel](https://user-images.githubusercontent.com/26483208/96576070-e555c380-12d1-11eb-8140-10f6a22b9750.gif)

JKMarkowski commented:
Multiple items bug
![carouselgfi](https://user-images.githubusercontent.com/26483208/96576086-ec7cd180-12d1-11eb-8046-d3b41bff3572.gif)

JKMarkowski commented on libs/core/src/lib/carousel/carousel.component.ts:
```diff
@@ -0,0 +1,412 @@
+    overflow = 'auto';
+
+    /** Is carousel is vertical. Default value is false. */
+    @Input()
+    vertical = false;
+
+    /** Number of items to be visible at a time */
+    @Input()
+    visibleSlidesCount = 1;
+
+    /** An event that is emitted after a slide transition has happened */
+    @Output()
+    readonly slideChange: EventEmitter<CarouselActiveSlides> = new EventEmitter<CarouselActiveSlides>();
```
Could you put API properties in valid places?

DeepakSap14 replied:
fixed

JKMarkowski commented:
Hello @DeepakSap14 Rest LGTM. It's almost ready to be merged, but those 2 bugs need to be addressed

DeepakSap14 commented:
> Multiple items bug
> ![carouselgfi](https://user-images.githubusercontent.com/26483208/96576086-ec7cd180-12d1-11eb-8046-d3b41bff3572.gif)

This is fixed now.

DeepakSap14 commented:
> IE11
> ![ie11carousel](https://user-images.githubusercontent.com/26483208/96576070-e555c380-12d1-11eb-8140-10f6a22b9750.gif)

currently do not have IE11 access. will check and fix in some time.
this should not be happening. May be image did not load completely. please verify once.

DeepakSap14 commented on apps/docs/src/app/core/component-docs/carousel/examples/carousel-looped-navigation.component.ts:
```diff
@@ -0,0 +1,9 @@
+    selector: 'fd-carousel-looped-navigation',
+    templateUrl: './carousel-looped-navigation.component.html',
+    changeDetection: ChangeDetectionStrategy.OnPush,
+    encapsulation: ViewEncapsulation.None
```
fixed

DeepakSap14 commented on apps/docs/src/app/core/component-docs/carousel/carousel-docs.component.html:
```diff
@@ -0,0 +1,96 @@
+<component-example>
+    <fd-carousel-vertical-direction></fd-carousel-vertical-direction>
+</component-example>
+<code-example [exampleFiles]="carouselVertical"></code-example>
```
had wrong initial height for carousel. fixed now.

DeepakSap14 commented:
> IE11
> ![ie11carousel](https://user-images.githubusercontent.com/26483208/96576070-e555c380-12d1-11eb-8140-10f6a22b9750.gif)

Looking into this.

DeepakSap14 commented:
> > IE11
> > ![ie11carousel](https://user-images.githubusercontent.com/26483208/96576070-e555c380-12d1-11eb-8140-10f6a22b9750.gif)
> 
> Looking into this.

This is fixed now using max-width: 100%, width:auto, height:auto.

DeepakSap14 commented:
Branch conflict will be resolved once platform table fixes use of "KeyUtil.isKey". on latest master rebase build error coming from platform table component.

JKMarkowski reviewed (approved):
LGTM. It's ready for merge

droshev reviewed (changes requested):
Great work! Few things:
- it looks like there is a double border on the bottom part:
<img width="339" alt="Screen Shot 2020-10-21 at 9 26 36 PM" src="https://user-images.githubusercontent.com/4380815/96807576-efb4b200-13e4-11eb-998a-a001ba2c38ea.png">
- there is a weird effect when you try to drag images:
![carousel](https://user-images.githubusercontent.com/4380815/96807602-fd6a3780-13e4-11eb-914a-ade450fc90bf.gif)
- can we achieve this sliding with the mouse from the carousel with multiple items in the carousel with 1 active item?
![sliding-carousel](https://user-images.githubusercontent.com/4380815/96807674-2ee30300-13e5-11eb-8c75-9fb80dcc15c1.gif)

InnaAtanasova commented:
Hey @DeepakSap14 ,
I found some minor issues testing the component on a mobile device: 
- [ ] if you slide slowly sometimes two images are visible at the same time. I understand it's a performance issue but if it's not addressed now could you create an issue for targeting this in the future. Here's a screenshot of the issue:
![Image from iOS (3)](https://user-images.githubusercontent.com/39598672/96912252-0eac5600-1470-11eb-903f-c5af571b971a.png)

- [ ] Landscape view, you can't make the last card appear in the view. The right button is disabled and the last "dot" is marked as active but you basically can't display the last item. See screenshot:
![Image from iOS (2)](https://user-images.githubusercontent.com/39598672/96912466-6f3b9300-1470-11eb-92dd-6d1cfab6c9b3.png)

- [ ] this is similar to the issue above. In portrait mode you can't display cards 7th and 8th:
![Image from iOS (1)](https://user-images.githubusercontent.com/39598672/96912601-9d20d780-1470-11eb-871f-3ea60670ab2f.png)

- [ ] Message page is not centered on a mobile device
![Image from iOS](https://user-images.githubusercontent.com/39598672/96912670-b3c72e80-1470-11eb-9f34-955a4ed97ad9.png)

DeepakSap14 commented:
> Great work! Few things:
> 
> * it looks like there is a double border on the bottom part:
> 
> <img alt="Screen Shot 2020-10-21 at 9 26 36 PM" width="339" src="https://user-images.githubusercontent.com/4380815/96807576-efb4b200-13e4-11eb-998a-a001ba2c38ea.png">
> 
> * there is a weird effect when you try to drag images:
>   ![carousel](https://user-images.githubusercontent.com/4380815/96807602-fd6a3780-13e4-11eb-914a-ade450fc90bf.gif)
> * can we achieve this sliding with the mouse from the carousel with multiple items in the carousel with 1 active item?
>   ![sliding-carousel](https://user-images.githubusercontent.com/4380815/96807674-2ee30300-13e5-11eb-8c75-9fb80dcc15c1.gif)

1. carousel styles does not define focus. So each browser was behaving differently by default. Now for documentation i have included, fd-carousel:focus styles. It is consistent now.
2. fixed by pointer-event: none style.
3. fixed. swipe working for single active item as well.

DeepakSap14 commented:
> Hey @DeepakSap14 ,
> I found some minor issues testing the component on a mobile device:
> 
> * [ ]  if you slide slowly sometimes two images are visible at the same time. I understand it's a performance issue but if it's not addressed now could you create an issue for targeting this in the future. Here's a screenshot of the issue:
>   ![Image from iOS (3)](https://user-images.githubusercontent.com/39598672/96912252-0eac5600-1470-11eb-903f-c5af571b971a.png)
> * [ ]  Landscape view, you can't make the last card appear in the view. The right button is disabled and the last "dot" is marked as active but you basically can't display the last item. See screenshot:
>   ![Image from iOS (2)](https://user-images.githubusercontent.com/39598672/96912466-6f3b9300-1470-11eb-92dd-6d1cfab6c9b3.png)
> * [ ]  this is similar to the issue above. In portrait mode you can't display cards 7th and 8th:
>   ![Image from iOS (1)](https://user-images.githubusercontent.com/39598672/96912601-9d20d780-1470-11eb-871f-3ea60670ab2f.png)
> * [ ]  Message page is not centered on a mobile device
>   ![Image from iOS](https://user-images.githubusercontent.com/39598672/96912670-b3c72e80-1470-11eb-9f34-955a4ed97ad9.png)

@InnaAtanasova 
1. I did not get this issue. When animation effect is sliding. then in this current active item slides out and new item slides in. In this process, until sliding is completed. part of both item will be visible. see this example: http://css3.bradshawenterprises.com/sliding/

2. fixed hidden card issue in Mobile view portrait and landscape.
3. fixed hidden card issue in Mobile view portrait and landscape.
4. Error message is center aligned in every screen now.

DeepakSap14 commented:
@InnaAtanasova @droshev This can be reviewed now.
I will hold this for merge as you had suggested.

InnaAtanasova reviewed (approved):
@DeepakSap14 I verified that my comments have been addressed. Thanks for doing this. Also, great job with the component.

droshev reviewed (changes requested):
![carousel](https://user-images.githubusercontent.com/4380815/97228825-d2e5f900-17ad-11eb-8c72-0481adff332c.gif)
looks much better!
One thing i noticed when you Tab over the carousel with multiple items the content gets shifted but the little dots indicating the current page don't

manavs commented:
> ![carousel](https://user-images.githubusercontent.com/4380815/97228825-d2e5f900-17ad-11eb-8c72-0481adff332c.gif)
> looks much better!
> One thing i noticed when you Tab over the carousel with multiple items the content gets shifted but the little dots indicating the current page don't.

May be I am viewing this incorrectly, but looks like when tabbing through to the card which is current is always the middle one. Not sure whether that it part of the requirement spec. This happens when the new set of cards come under view

DeepakSap14 commented:
> ![carousel](https://user-images.githubusercontent.com/4380815/97228825-d2e5f900-17ad-11eb-8c72-0481adff332c.gif)
> looks much better!
> One thing i noticed when you Tab over the carousel with multiple items the content gets shifted but the little dots indicating the current page don't.

@droshev 
taborder issue has been fixed now.
please review.

droshev reviewed (commented):
@DeepakSap14 Great work! LGTM 🚀
